pip/pip-285.md

```diff
@@ -0,0 +1,73 @@
+
+The configuration and alerting settings for this monitoring item play a crucial role in responding swiftly to issues. When "pulsar_subscription_backlog_duration" indicates an abnormal increase in duration or unusual message backlogs, system administrators receive immediate alert notifications. These alerts enable administrators to quickly identify problems and take necessary measures to prevent message losses or further delays.
+
+In conclusion, the introduction of the "pulsar_subscription_backlog_duration" monitoring item enables effective monitoring of subscription health, real-time issue detection, and prevention of message delivery delays and losses. Additionally, timely alerting mechanisms empower proactive responses, ensuring the reliability and efficiency of the messaging system. This is essential for providing high-quality message delivery services, ensuring user experiences, and maintaining data integrity.
+
+# Goals
+
+## In Scope
+
+* SubscriptionStatsImpl add this stat
+* Metrics
+
+
+## Out of Scope
+
+* Implementing changes to the core functionality of the Pulsar messaging system itself.
+* Not include `NonPersistentTopic`.
+* Not include `DelayMessage`
+
+# High Level Design
+
+* add config `subscriptionBacklogDurationEnabled` in `broker.conf`
+  * note: because we need to read the markDelete position next position message, it will consume performance when the message is not in the cache, so add this flag
+* `SubscriptionStatsImpl` add `backlogDuration` variable
+* `AggregatedSubscriptionStats` add `backlogDuration` variable
+* add metric iterm named `pulsar_subscription_back_log_duration`
+
+# Detailed Design
+
+
+## Design & Implementation Details
+* when `PersistentSubscription` invoke getStats then reade the (`markDelete` + 1) entry convert to `MessageMetadata` `publish_time` to represent the `earliestUnAckMessagePublishTime`
+* use currentTime - `earliestUnAckMessagePublishTime` represent `backlogDuration`
+* if markDelete haven't changed, don't need to get the new `earliestUnAckMessagePublishTime`, use directly, to reduce the read entry op
+
+## Public-facing Changes
+
+
+### Configuration
+
+* add config `subscriptionBacklogDurationEnabled` in `broker.conf`
+  * note: because we need to read the markDelete position next position message, it will consume performance when the message is not in the cache, so add the config
+
+# Backward & Forward Compatability
+
+## Revert
+
+* config `subscriptionBacklogDurationEnabled = false` in `broker.conf`
+* lower the broker version
+
+## Upgrade
+
+* config `subscriptionBacklogDurationEnabled = true` in `broker.conf`
+
+# Alternatives
+
+* marDeletePosition changed every time change the `earliestUnAckMessagePublishTime`, It will be very frequent and consume performance
```

Comment:
You mean to say:
Everytime markDeletePosition changes, we'll read its message and keep in memory it's timestatmp. When the metrics endpoint is invoked, we'll simpy read it from memory. The cost of this alternative is very high since in worst case we'll read all messages, but backwards which means they will likely not even be in cache.

Reply:
yes, so it's an alternative, for now I can't think of anything better

Reply:
My comment a different way to phrase what you wrote. Not arguing about it, since I agree. I just want other readers to understand it better, hence offered a better way to write it.

Regarding a more efficient way - I offered one in other comment.